syntax = "proto3";
package obsremote;

// 視頻編碼類型
enum VideoCodec {
  CODEC_RAW_BGRA = 0;  // 無壓縮 BGRA
  CODEC_H264 = 2;      // H.264/NVENC
}

// 屬性類型 (與 OBS obs_property_type 對應，只列出實際支援的類型)
enum PropertyType {
  PROPERTY_UNKNOWN = 0;
  PROPERTY_BOOL = 1;
  PROPERTY_INT = 2;
  PROPERTY_FLOAT = 3;
  PROPERTY_TEXT = 4;
  PROPERTY_PATH = 5;
  PROPERTY_LIST = 6;
  PROPERTY_COLOR = 7;
  PROPERTY_BUTTON = 8;        // 不支援傳輸，但保留可識別
  PROPERTY_COLOR_ALPHA = 13;  // OBS 原生值
}

// 數字類型 (slider vs spinbox)
enum NumberType {
  NUMBER_SCROLLER = 0;
  NUMBER_SLIDER = 1;
}

// 文字類型 (obs_text_type)
enum TextType {
  TEXT_DEFAULT = 0;
  TEXT_PASSWORD = 1;
  TEXT_MULTILINE = 2;
  TEXT_INFO = 3;
}

// 文字資訊類型 (obs_text_info_type)
enum TextInfoType {
  TEXT_INFO_NORMAL = 0;
  TEXT_INFO_WARNING = 1;
  TEXT_INFO_ERROR = 2;
}

// 路徑類型
enum PathType {
  PATH_FILE = 0;
  PATH_FILE_SAVE = 1;
  PATH_DIRECTORY = 2;
}

// 列表格式
enum ListFormat {
  LIST_FORMAT_STRING = 0;
  LIST_FORMAT_INT = 1;
  LIST_FORMAT_FLOAT = 2;
}

// 列表項
message ListItem {
  string name = 1;
  string value_string = 2;
  int64 value_int = 3;
  double value_float = 4;
}

// 屬性定義
message Property {
  string name = 1;
  string description = 2;
  PropertyType type = 3;
  bool visible = 4;
  bool enabled = 5;
  string long_description = 6;
  
  // LIST 類型
  repeated ListItem items = 10;
  string current_string = 11;
  int64 current_int = 12;
  double current_float = 13;
  ListFormat list_format = 14;
  
  // BOOL 類型
  bool current_bool = 20;
  
  // INT 類型
  int32 min_int = 30;
  int32 max_int = 31;
  int32 step_int = 32;
  int32 current_int_value = 33;
  NumberType int_type = 34;
  
  // FLOAT 類型
  double min_float = 40;
  double max_float = 41;
  double step_float = 42;
  double current_float_value = 43;
  NumberType float_type = 44;
  
  // TEXT 類型
  string current_text = 50;
  TextType text_type = 51;
  TextInfoType text_info_type = 52;
  
  // PATH 類型
  PathType path_type = 60;
  string path_filter = 61;
  string path_default = 62;
}

// ========== Session-based API ==========

// 空消息
message Empty {}

// Source 資訊
message SourceInfo {
  string id = 1;            // e.g., "window_capture"
  string display_name = 2;  // e.g., "視窗擷取"
}

// 獲取可用 source 類型
message AvailableSourcesResponse {
  repeated SourceInfo sources = 1;
}

// 創建 Session 響應
message CreateSessionResponse {
  string session_id = 1;
}

// 設定 Source Type
message SetSourceTypeRequest {
  string session_id = 1;
  string source_type = 2;
}

message SetSourceTypeResponse {
  repeated Property properties = 1;
}

// 更新設定
message UpdateSettingsRequest {
  string session_id = 1;
  string settings_json = 2; // JSON string of obs_data_t
}

message UpdateSettingsResponse {
  repeated Property properties = 1;
}

// 獲取屬性 (Session 版本)
message GetPropertiesRequest {
  string session_id = 1;
}

message GetPropertiesResponse {
  repeated Property properties = 1;
}

// 開始串流 (Session 版本)
message StartStreamRequest {
  string session_id = 1;
}

// 釋放 Session
message ReleaseSessionRequest {
  string session_id = 1;
}

// 音頻狀態查詢
message IsAudioActiveRequest {
  string session_id = 1;
}

message IsAudioActiveResponse {
  bool audio_active = 1;
}

// SRT 串流資訊查詢
message GetSrtInfoRequest {
  string session_id = 1;
}

message SrtInfo {
  int32 port = 1;           // SRT 連接 Port
  int32 latency_ms = 2;     // 延遲設定 (ms)
  string passphrase = 3;    // 加密密碼 (可選)
  bool ready = 4;           // 是否就緒
}

// ========== 時鐘同步 ==========

// 時鐘同步請求 (Client 發送自己的時間戳)
message SyncClockRequest {
  int64 client_send_time_ns = 1;  // Client 發送時的 os_gettime_ns()
}

// 時鐘同步響應 (Server 回應自己的時間戳)
message SyncClockResponse {
  int64 server_time_ns = 1;       // Server 收到請求時的 os_gettime_ns()
}

// ========== 串流數據 ==========

// 視頻幀
message VideoFrame {
  uint32 width = 1;
  uint32 height = 2;
  VideoCodec codec = 3;
  bytes frame_data = 4;
  uint64 timestamp_ns = 5;
  uint32 frame_number = 6;
  uint32 linesize = 7;
}

// 音頻幀
message AudioFrame {
  uint32 sample_rate = 1;
  uint32 channels = 2;
  bytes frame_data = 3;  // 原始音頻數據 (planar float)
  uint64 timestamp_ns = 4;
}

// 串流開始資訊 (在第一幀前發送)
message StreamInfo {
  int64 stream_start_time_ns = 1;  // Server 端 os_gettime_ns() 記錄的串流開始時間
}

// 串流幀 (影音/控制訊息)
message StreamFrame {
  oneof data {
    VideoFrame video = 1;
    AudioFrame audio = 2;
    StreamInfo stream_info = 3;  // 串流開始時發送一次
  }
}

// ========== 服務定義 ==========
service RemoteCaptureService {
  // 獲取可用的 source 類型列表 (由 server 決定)
  rpc GetAvailableSources(Empty) returns (AvailableSourcesResponse);
  
  // Session 管理
  rpc CreateSession(Empty) returns (CreateSessionResponse);
  rpc ReleaseSession(ReleaseSessionRequest) returns (Empty);
  
  // Source 操作 (在 session 內)
  rpc SetSourceType(SetSourceTypeRequest) returns (SetSourceTypeResponse);
  rpc UpdateSettings(UpdateSettingsRequest) returns (UpdateSettingsResponse);
  rpc GetProperties(GetPropertiesRequest) returns (GetPropertiesResponse);
  
  // 串流 (綁定 session_id)
  rpc StartStream(StartStreamRequest) returns (stream StreamFrame);
  
  // 音頻狀態查詢 (延遲調用)
  rpc IsAudioActive(IsAudioActiveRequest) returns (IsAudioActiveResponse);
  
  // SRT 串流資訊
  rpc GetSrtInfo(GetSrtInfoRequest) returns (SrtInfo);
  
  // 時鐘同步 (RTT 測量)
  rpc SyncClock(SyncClockRequest) returns (SyncClockResponse);
}
